src-java/kilda-persistence-tinkerpop/src/test/java/org/openkilda/persistence/ferma/repositories/FermaHaFlowRepositoryTest.java

```diff
@@ -0,0 +1,301 @@
+        assertPathsFlows(foundFlow1.get().getPaths(), path1, path2);
+
+        // flow 2
+        Optional<HaFlow> foundFlow2 = haFlowRepository.findById(HA_FLOW_ID_2);
+        assertTrue(foundFlow2.isPresent());
+        assertHaFlow(HA_FLOW_ID_2, SWITCH_ID_2, PORT_2, VLAN_2, INNER_VLAN_2, LATENCY_3, LATENCY_4, BANDWIDTH_2,
+                FlowEncapsulationType.VXLAN, PRIORITY_2, DESCRIPTION_2, PathComputationStrategy.LATENCY,
+                FlowStatus.IN_PROGRESS, false, false, false, false, false, foundFlow2.get());
+        assertEquals(0, foundFlow2.get().getSubFlows().size());
+        assertEquals(0, foundFlow2.get().getPaths().size());
+    }
+
+    @Test
+    public void findByIdNoExistentFlowTest() {
+        assertFalse(haFlowRepository.findById("123").isPresent());
+    }
+
+    @Test
+    public void haFlowCascadeRemoveTest() {
+        createHaFlow(haFlow1);
+        createHaFlow(haFlow2);
+        haSubFlowRepository.add(subFlow1);
+        haSubFlowRepository.add(subFlow2);
+        haFlow1.setSubFlows(Sets.newHashSet(subFlow1, subFlow2));
+
+        haFlowPathRepository.add(path1);
+        path1.setSegments(buildSegments(path1, switch1, switch2, switch3));
+        path1.setHaSubFlowEdges(buildSubFlowEdges(subFlow1, subFlow2));
+
+        haFlowPathRepository.add(path2);
+        path2.setSegments(buildSegments(path2, switch3, switch2, switch1));
+        path2.setHaSubFlowEdges(buildSubFlowEdges(subFlow2, subFlow1));
+        haFlow1.addPaths(path1, path2);
+
+        assertEquals(2, haFlowRepository.findAll().size());
+        assertEquals(2, haFlowPathRepository.findAll().size());
+        assertEquals(2, haSubFlowRepository.findAll().size());
+        assertEquals(2, pathSegmentRepository.findByPathId(path1.getHaPathId()).size());
+        assertEquals(2, pathSegmentRepository.findByPathId(path2.getHaPathId()).size());
+
+        Optional<HaFlow> removedFlow = haFlowRepository.remove(haFlow1.getHaFlowId());
+        assertTrue(removedFlow.isPresent());
+        assertHaFlow(HA_FLOW_ID_1, SWITCH_ID_1, PORT_1, VLAN_1, INNER_VLAN_1, LATENCY_1, LATENCY_2, BANDWIDTH_1,
+                FlowEncapsulationType.TRANSIT_VLAN, PRIORITY_1, DESCRIPTION_1, PathComputationStrategy.COST,
+                FlowStatus.UP, true, true, true, true, true, removedFlow.get());
+
+        assertTrue(haFlowRepository.findById(haFlow2.getHaFlowId()).isPresent());
+        assertEquals(1, haFlowRepository.findAll().size());
+        assertEquals(0, haFlowPathRepository.findAll().size());
+        assertEquals(0, haSubFlowRepository.findAll().size());
+        assertEquals(0, pathSegmentRepository.findByPathId(path1.getHaPathId()).size());
+        assertEquals(0, pathSegmentRepository.findByPathId(path2.getHaPathId()).size());
+    }
+
+    private void assertPathsFlows(Collection<HaFlowPath> actualPaths, HaFlowPath... expectedPaths) {
+        Map<PathId, HaFlowPath> actualMap = haFlowPathsToMap(actualPaths);
```

Comment:
This assertion doesn't take into consideration redundant paths. You assert the size in `createHaFlowTest`, but don't assert in `findByIdHaFlowTest`. I think you can simplify it by converting the input collection to a set, as well as put the expected vars into set, then use assert equals on sets. 
A separate question is about uniqueness: actualPaths can have dupes, but they disappear once it is converted to the map.

Reply:
thanks! in fact, equals for sets implies they have the same size, otherwise at least one element is not present in other set. So with the sets approach line 280 is redundant.